File: antlir/bzl/tests/shapes/test_shape.py
#!/usr/bin/env python3
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

import importlib
import unittest
from typing import Mapping, Optional, Sequence, Tuple

from antlir.bzl.target import target_t

# TODO remove all references to hashable and just use characters once
# read-only dicts land
from antlir.bzl.tests.shapes.test import (
    character_collection_t,
    character_t,
    friend_t,
    hashable_t,
)

from antlir.freeze import frozendict
from antlir.fs_utils import Path
from antlir.shape import Shape


# pyre-fixme[16]: `character_t` has no attribute `types`.
lightsaber_t = character_t.types.weapon.__args__[0]
characters = character_collection_t.from_env("characters").characters

# @oss-disable[end= ]: TARGET_PATH = "fbcode//antlir/bzl/tests/shapes"
TARGET_PATH = "antlir//antlir/bzl/tests/shapes" # @oss-enable


class TestShape(unittest.TestCase):
    def setUp(self):
        # More output for easier debugging
        unittest.util._MAX_LENGTH = 12345
        self.maxDiff = 12345

    def test_load(self):
        c = characters[0]
        self.assertEqual(c.name, "Luke Skywalker")
        self.assertEqual(c.appears_in, (4, 5, 6))
        self.assertEqual(
            [f.name for f in c.friends], ["Han Solo", "Leia Organa", "C-3PO"]
        )
        self.assertEqual(
            c.weapon,
            lightsaber_t(
                color=lightsaber_t.types.color.GREEN,
                target=target_t(
                    name=f"{TARGET_PATH}:luke-lightsaber",
                    path=b"",
                ),
            ),
        )
        self.assertEqual(c.metadata, frozendict({"species": "human"}))
        self.assertEqual(c.personnel_file, Path("/rebellion/luke_skywalker.txt"))
        self.assertIsInstance(c.personnel_file, Path)

    def test_data_and_resources(self):
        # hashable_t is just a (incomplete) subset of the character type that
        # is both hashable and serializable to python_data/json_file
        res = hashable_t.read_resource(__package__, "data.json")
        self.assertEqual(res.name, "Stormtrooper")
        # load the same thing from a file path
        with importlib.resources.path(__package__, "data.json") as path:
            f = hashable_t.load(path)
        self.assertEqual(f, res)
        # lastly, the directly imported python_data version should also be
        # equivalent
        from antlir.bzl.tests.shapes.data import data as imp

        self.assertEqual(imp, res)
        self.assertTrue(isinstance(imp, hashable_t))

    def test_json_file_targets(self):
        with importlib.resources.path(__package__, "characters.json") as path:
            c = character_collection_t.load(path)
        target = c.characters[0].weapon.target
        self.assertEqual(target.name, f"{TARGET_PATH}:luke-lightsaber")

    def test_hash(self):
        trooper1 = hashable_t(
            name="Stormtrooper",
            appears_in=[1, 2, 3, 4, 5, 6],
            friends=[],
        )
        trooper2 = hashable_t(
            name="Stormtrooper",
            appears_in=[1, 2, 3, 4, 5, 6],
            friends=[],
        )
        self.assertEqual(trooper1.__hash__(), trooper2.__hash__())

    def test_typehints(self):
        """check type hints on generated classes"""

        def _deep_typehints(obj):
            hints = {}
            for key, val in getattr(obj, "__annotations__", {}).items():
                if hasattr(val, "__annotations__"):
                    hints[key] = _deep_typehints(val)
                else:
                    hints[key] = val
            return hints

        # non-nested type hints can easily be compared
        hints = _deep_typehints(character_t)
        expected = {
            "name": str,
            "appears_in": Tuple[int, ...],
            "metadata": Mapping[str, str],
        }
        self.assertEqual(expected, {k: v for k, v in hints.items() if k in expected})

    def test_instance_repr(self):
        # Shapes don't always have nice classnames, so the repr is customized to
        # be human-readable. While this has no functional impact on the code, it
        # is critical for usability, so ensure there are unit tests.
        self.assertEqual(
            repr(characters[0]),
            "shape("
            "affiliations=shape(faction='Rebellion'), "
            "appears_in=(4, 5, 6), "
            "friends=("
            + (
                "shape(name='Han Solo'), "
                "shape(name='Leia Organa'), "
                "shape(name='C-3PO')"
            )
            + "), "
            "metadata=frozendict({'species': 'human'}), "
            "name='Luke Skywalker', "
            "personnel_file=b'/rebellion/luke_skywalker.txt', "
            "weapon=shape("
            + (
                "color=GREEN, "
                "target=shape("
                f"name='{TARGET_PATH}:luke-lightsaber', "
                "path=b''"
                ")"
            )
            + "))",
        )

    def test_class_repr(self):
        # The generated classes also have a custom repr, which is much more
        # readable. However, it's a huge pain to test that it actually looks
        # good, so just make sure it doesn't look like the default python class
        # repr
        self.assertNotIn("<class", repr(character_t))
        self.assertTrue(repr(character_t).startswith("shape("))

    def test_immutable_fields(self):
        with self.assertRaises(TypeError):
            characters[0].name = "Darth Vader's son"
        # immutability is a property at any level of nesting
        with self.assertRaises(TypeError):
            characters[0].friends[0].name = "R2-D2"
        # shouldn't be able to add to any lists
        new_friend = friend_t(name="R2-D2")
        with self.assertRaises(AttributeError):
            characters[0].friends.append(new_friend)
        # or dictionaries
        with self.assertRaises(TypeError):
            characters[0].metadata["favorite-food"] = "Blue Milk"

    def test_subclass(self):
        """
        Demonstrate a pure Python subclass of a shape type with added fields
        and functions.
        """

        class Jedi(character_t):
            padawan: Optional[str]

            def train(self) -> str:
                if self.padawan:
                    return f"Training {self.padawan}"
                return "No one to train right now"

        obi_wan = Jedi(
            padawan="Anakin Skywalker",
            name="Obi-Wan Kenobi",
            appears_in=[1, 2, 3, 4, 5, 6],
            # TODO: there is not a good way to get at the type of an inner
            # shape definition. It's not super important at this point since
            # they should really only be deserialized from JSON that comes from
            # Buck macros, and in the rare case that it's necessary, a dict
            # with the same fields works and is properly validated.
            friends=[{"name": "Yoda"}, {"name": "Padme Amidala"}],
            weapon=lightsaber_t(color="blue"),
            affiliations=character_t.types.affiliations(faction="Jedi Temple"),
        )
        # subclass should still be immutable by default
        with self.assertRaises(TypeError):
            obi_wan.padawan = "Luke Skywalker"

        # subclass instance repr uses the human-written class name, and the
        # inner shapes use shape()
        self.assertEqual(
            repr(obi_wan),
            "Jedi("
            "affiliations=shape(faction='Jedi Temple'), "
            "appears_in=(1, 2, 3, 4, 5, 6), "
            "friends=(shape(name='Yoda'), shape(name='Padme Amidala')), "
            "metadata=frozendict({'species': 'human'}), "
            "name='Obi-Wan Kenobi', "
            "personnel_file=None, "
            "weapon=shape(color=BLUE, target=None), "
            "padawan='Anakin Skywalker'"
            ")",
        )
        # subclass type repr should also use the human-written class name
        self.assertTrue(repr(Jedi).startswith("Jedi("))
        # added field shows up
        self.assertIn("padawan=", repr(Jedi))

        self.assertEqual(obi_wan.train(), "Training Anakin Skywalker")

    def test_nested_shape_class(self):
        self.assertEqual(
            character_t.types.affiliations(faction="shape.bzl").faction,
            "shape.bzl",
        )
        # collection fields should do the sane thing
        # dicts go to value type
        self.assertEqual(character_t.types.metadata, str)
        # lists go to the homogenous element type
        self.assertEqual(
            character_t.types.friends,
            character_t.__annotations__["friends"].__args__[0],
        )

    def test_default_shape(self):
        """default values for nested shapes must be deserialized"""
        c3po = characters[2]
        self.assertEqual(c3po.name, "C-3PO")
        self.assertEqual(c3po.affiliations.faction, "Rebellion")

    def test_failure_when_has_types_field(self):
        with self.assertRaises(KeyError):

            class AnnotationOnly(Shape):
                # fake that this came from a class defined by shape.bzl
                __GENERATED_SHAPE__ = True
                # this is reserved for the type definitions class
                types: Sequence[str]

        with self.assertRaises(KeyError):

            class WithDefault(Shape):
                # fake that this came from a class defined by shape.bzl
                __GENERATED_SHAPE__ = True
                # this is reserved for the type definitions class
                types: Sequence[str] = ("hello", "world")

    def test_rendered_template(self):
        self.assertEqual(
            "Stormtrooper is a character that appears in episode(s) 1, 2, 3, 4, 5 and 6 of\n"
            "the Star Wars franchise and is friends with Vader, Palpatine and Tarkin.",
            importlib.resources.read_text(__package__, "template.txt"),
        )

File: antlir/bzl/tests/shapes/test_shape.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//! These tests don't have to be anywhere near as exhaustive as those of
//! test_shape.py, because rust gives so much type-safety already. Basically, as
//! long as `serde_json::from_str` is able to load the `character_collection_t`,
//! everything else that can be done with shapes are safe.
use anyhow::Context;
use anyhow::Result;
use shape::ShapePath;
use target::target_t;
use test_shape::character_collection_t;
use test_shape::character_t;
use test_shape::friend_t;
use test_shape::inner;
use test_shape::target_with_path_t;
use test_shape::thrift_new;
use test_shape::thrift_old;
use test_shape::union_new;
use test_shape::union_old;
use test_shape::with_default_trait;
use test_shape::with_optional_int;

// @oss-disable: const TARGET_PREFIX: &str = "fbcode//antlir/bzl/tests/shapes";
const TARGET_PREFIX: &str = "antlir//antlir/bzl/tests/shapes"; // @oss-enable

#[test]
fn load() -> Result<()> {
    let characters_str = std::env::var("characters").context("missing 'characters' env var")?;
    let mut characters: character_collection_t = serde_json::from_str(&characters_str)
        .with_context(|| format!("failed to parse json '{}'", characters_str))?;
    let luke = characters.characters.remove(0);
    assert_eq!(luke.name, "Luke Skywalker");
    assert_eq!(luke.appears_in, [4, 5, 6]);
    assert_eq!(
        luke.friends,
        ["Han Solo", "Leia Organa", "C-3PO"]
            .into_iter()
            .map(|n| friend_t { name: n.into() })
            .collect::<Vec<_>>()
    );
    Ok(())
}

#[test]
fn defaults() -> Result<()> {
    // this really checks that the default values are deserialized from a shape
    // input (well-structured from the buck graph)
    let characters_str = std::env::var("characters").context("missing 'characters' env var")?;
    let mut characters: character_collection_t = serde_json::from_str(&characters_str)
        .with_context(|| format!("failed to parse json '{}'", characters_str))?;
    let c3po = characters.characters.remove(2);
    assert_eq!(c3po.name, "C-3PO");
    assert_eq!(c3po.affiliations.faction, "Rebellion");

    // check that the default values are populated even with empty json
    let c3po: character_t =
        serde_json::from_str(r#"{"name": "C-3PO", "appears_in": [1,2,3,4,5,6], "friends": []}"#)
            .context("failed to deserialize json without explicit default values")?;
    assert_eq!(c3po.name, "C-3PO");
    assert_eq!(c3po.affiliations.faction, "Rebellion");

    Ok(())
}

#[test]
fn thrift() -> Result<()> {
    let characters_str = std::env::var("characters").context("missing 'characters' env var")?;
    let mut characters: character_collection_t = serde_json::from_str(&characters_str)
        .with_context(|| format!("failed to parse json '{}'", characters_str))?;
    let luke = characters.characters.remove(0);
    let after_roundtrip =
        fbthrift::binary_protocol::deserialize(fbthrift::binary_protocol::serialize(&luke))?;
    assert_eq!(luke, after_roundtrip);

    assert_eq!(
        serde_json::json!({
            "name":"Luke Skywalker",
            "metadata":{"species":"human"},
            "affiliations": {
                "faction":"Rebellion"
            },
            "appears_in":[4,5,6],
            "friends":[
                {"name":"Han Solo"},
                {"name":"Leia Organa"},
                {"name":"C-3PO"},
            ],
            "personnel_file":"/rebellion/luke_skywalker.txt",
            "weapon":{
                "lightsaber_t":{
                    "color": "green",
                    "target":{
                        "name": format!("{TARGET_PREFIX}:luke-lightsaber"),
                        "path": "",
                    }
                }
            }
        }),
        serde_json::from_slice::<serde_json::Value>(&fbthrift::simplejson_protocol::serialize(
            &luke
        ))
        .expect("failed to round trip through thrift")
    );

    Ok(())
}

#[test]
fn thrift_compat() {
    assert_eq!(
        thrift_new {
            foo: 42,
            baz: Some("baz".into()),
            qux: None
        },
        fbthrift::binary_protocol::deserialize(fbthrift::binary_protocol::serialize(thrift_old {
            foo: 42,
            bar: Some("hello-bar".into()),
        }))
        .expect("could not deserialize thrift_old as thrift_new")
    );
    assert_eq!(
        thrift_old { foo: 42, bar: None },
        fbthrift::binary_protocol::deserialize(fbthrift::binary_protocol::serialize(thrift_new {
            foo: 42,
            baz: Some("hello-baz".into()),
            qux: Some(true),
        }))
        .expect("could not deserialize thrift_new as thrift_old")
    );
    assert_eq!(
        union_old::Int(42),
        fbthrift::binary_protocol::deserialize(fbthrift::binary_protocol::serialize(
            union_new::Int(42)
        ))
        .expect("could not deserialize union_new as union_old"),
    );
    assert_eq!(
        union_new::Int(42),
        fbthrift::binary_protocol::deserialize(fbthrift::binary_protocol::serialize(
            union_old::Int(42)
        ))
        .expect("could not deserialize union_old as union_new"),
    );
}

#[test]
fn optional_int() -> Result<()> {
    let s: with_optional_int = serde_json::from_str(r#"{"optint": null}"#)?;
    assert_eq!(s, with_optional_int { optint: None });
    let s: with_optional_int = serde_json::from_str(r#"{"optint": 42}"#)?;
    assert_eq!(s, with_optional_int { optint: Some(42) });
    let s: with_optional_int =
        serde_json::from_str(r#"{"foo": "bar"}"#).context("failed to deser")?;
    assert_eq!(s, with_optional_int { optint: None });
    Ok(())
}

#[test]
fn default_trait() -> Result<()> {
    let with_default: with_default_trait = Default::default();

    assert_eq!(
        with_default,
        with_default_trait {
            a: Some("abc".to_string()),
            b: true,
            c: inner {
                a: Some("def".to_string())
            },
            d: None,
        }
    );

    Ok(())
}

#[test]
fn default_builder() -> Result<()> {
    let with_default = with_default_trait::builder()
        .a(Some("test".to_string()))
        .build();

    assert_eq!(
        with_default,
        with_default_trait {
            a: Some("test".to_string()),
            b: true,
            c: inner {
                a: Some("def".to_string())
            },
            d: None,
        }
    );
    Ok(())
}

#[test]
fn target_forwards_compatible() {
    let new_from_old: target_t = fbthrift::binary_protocol::deserialize(
        fbthrift::binary_protocol::serialize(target_with_path_t {
            name: "foo//bar:baz".into(),
            path: ShapePath::new("lorem/ipsum/dolor".to_owned()),
        }),
    )
    .expect("failed to deserialize new_from_old");
    assert_eq!(
        new_from_old,
        target_t {
            name: "foo//bar:baz".into(),
            path: Some(ShapePath::new("lorem/ipsum/dolor".into()))
        }
    );
}

#[test]
fn target_backwards_compatible() {
    let old_from_new: target_with_path_t =
        fbthrift::binary_protocol::deserialize(fbthrift::binary_protocol::serialize(target_t {
            name: "foo//bar:baz".into(),
            path: Some(ShapePath::new("".to_owned())),
        }))
        .expect("failed to deserialize old_from_new");
    assert_eq!(
        old_from_new,
        target_with_path_t {
            name: "foo//bar:baz".into(),
            path: ShapePath::new("".to_owned()),
        }
    );
}
